### PR TITLE
Unsubscribe when close pulsar consumer

### PR DESCRIPTION
**Pull request by bigsheeper**
Signed-off-by: bigsheeper <yihao.dai@zilliz.com>

issues: #15193

### CODE REVIEW

File: internal/proxy/channels_mgr.go [hunk 1]
```diff
@@ -392,7 +392,10 @@ func (mgr *singleTypeChannelsMgr) createMsgStream(collectionID UniqueID) error {
 		stream.SetRepackFunc(mgr.repackFunc)
 	}
 	runtime.SetFinalizer(stream, func(stream msgstream.MsgStream) {
-		stream.Close()
+		err = stream.Close()
+		if err != nil {
+			log.Error("proxy singleTypeChannelsMgr close message stream failed", zap.Error(err))
```
Comment (xige-16) [inline internal/proxy/channels_mgr.go:397]: panic？  may cause channel be unable to unsubscribe

Comment (bigsheeper) [reply internal/proxy/channels_mgr.go]: fixed

File: internal/util/mqclient/pulsar_consumer.go [hunk 2]
```diff
@@ -105,7 +106,15 @@ func (pc *PulsarConsumer) Ack(message Message) {
 // Close the consumer and stop the broker to push more messages
 func (pc *PulsarConsumer) Close() {
 	pc.c.Close()
```
Comment (xige-16) [inline internal/util/mqclient/pulsar_consumer.go:108]: unsubscribe closed consumer will always failed, should unsub before close consumer

**Comment by xige-16** [comment]
rerun ut

File: internal/util/mqclient/pulsar_consumer.go [hunk 3]
```diff
@@ -34,6 +34,7 @@ type PulsarConsumer struct {
 	closeCh    chan struct{}
 	once       sync.Once
 	skip       bool
+	isClosed   bool
```
Comment (sunby) [inline internal/util/mqclient/pulsar_consumer.go:37]: If `Close` could be called by many goroutines at the same time. Please use atomic or sync.once

**Review by sunby — APPROVED** [review]

File: internal/util/mqclient/pulsar_consumer.go [hunk 4]
```diff
@@ -104,8 +105,16 @@ func (pc *PulsarConsumer) Ack(message Message) {
 
 // Close the consumer and stop the broker to push more messages
 func (pc *PulsarConsumer) Close() {
-	pc.c.Close()
-	close(pc.closeCh)
+	if !pc.isClosed {
+		// Unsubscribe for the consumer
+		err := pc.c.Unsubscribe()
+		if err != nil {
+			panic(err)
```
Comment (sunby) [inline internal/util/mqclient/pulsar_consumer.go:112]: Should the process panic if the `Unsubscribe` failed? I dont think the failure of `Unsubscribe` is severe like that.

File: internal/util/mqclient/pulsar_consumer_test.go [hunk 5]
```diff
@@ -55,3 +55,29 @@ func Test_PatchEarliestMessageID(t *testing.T) {
...
+	assert.NotNil(t, consumer)
+
+	str := consumer.Subscription()
+	assert.NotNil(t, str)
+
+	pulsarConsumer := &PulsarConsumer{c: consumer, closeCh: make(chan struct{})}
+	pulsarConsumer.Close()
+
+	// test double close
+	pulsarConsumer.Close()
```
Comment (sunby) [inline internal/util/mqclient/pulsar_consumer_test.go:82]: I think you should check something to verify that it succeed to close the consumer.

File: internal/proxy/channels_mgr.go (hunk 1 above)
Comment (sunby) [reply internal/proxy/channels_mgr.go]: > panic？ may cause channel be unable to unsubscribe

Will panic unsubcribe the channel?

File: internal/util/mqclient/pulsar_consumer.go (hunk 4 above)
Comment (bigsheeper) [reply internal/util/mqclient/pulsar_consumer.go:113]: panic and would retried by Coord

**Pull request merged**